### PR TITLE
Fixes #32489 - setting load on code reload

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Code reloading unloads setting descendants and thus
SettingRegistry.reload won't get any definitions of default from them.

This loads the STI classes on code reload.

Please let's get rid of the Setting STI.

### CODE REVIEW

File: config/initializers/foreman.rb [hunk 1]
```diff
@@ -42,8 +42,16 @@
 Foreman::Plugin.medium_providers_registry.register MediumProviders::Default
 
 Rails.application.config.to_prepare do
-  # Force reload settings after all plugins have loaded and on code reload
-  Foreman.settings.load if (Setting.table_exists? rescue(false))
+  if (Setting.table_exists? rescue(false))
+    # Force reload settings after all plugins have loaded and on code reload
+    Dir[
+      File.join(Rails.root, "app/models/setting.rb"),
+      File.join(Rails.root, "app/models/setting/*.rb"),
```
Comment (ekohl) [inline config/initializers/foreman.rb]: Not sure if it works (perhaps it needs `.to_s`) but isn't this cleaner, relying on the fact `Rails.root` is a [pathname](https://ruby-doc.org/stdlib-2.6.3/libdoc/pathname/rdoc/Pathname.html):
```suggestion
      Rails.root.join('app', 'models', 'setting.rb'),
      Rails.root.join('app', 'models', 'setting', '*.rb'),
```

I also see we have the same implementation around line 20. Is this a static method we should implement on `Settings`?

**Comment by lzap** [comment]
Nice, so according to https://guides.rubyonrails.org/configuring.html#initialization-events this block is called on every reload in devel mode.

File: config/initializers/foreman.rb (hunk 1 above)
Comment (ezr-ondrej) [reply config/initializers/foreman.rb]: I agree with slight improvement, if we are ok to change both, but no static method for this uglieness please.
We want to get rid of it by finishing the Setting dsl and migrating from STI to this DSL

**Comment by ezr-ondrej** [comment]
[test katello]

**Comment by ezr-ondrej** [comment]
[test katello]

**Comment by ezr-ondrej** [comment]
[test katello]

**Review by tbrisker — CHANGES_REQUESTED** [review]
This does fix reloading for core settings, but plugin settings are still not loaded properly on reload, leading to a ton of debug messages on every request.

**Comment by ezr-ondrej** [comment]
It is true, given it's a temporary fix, I can do a guessing game and go through all engines and include their paths.

**Comment by ezr-ondrej** [comment]
Guessed, I don't love this solution, but it works, so IMHO we should merge this, given its temporary fix

**Comment by lzap** [comment]
So what are we aiming for with this patch? Fixing reloading only for core, not plugins, but no debug messages? Or are you saying you think this is worth merging even with those messages?

**Comment by ezr-ondrej** [comment]
> So what are we aiming for with this patch? Fixing reloading only for core, not plugins, but no debug messages?

This is fixing reload for core and plugins now, messages should go away because it's fixed.
I hope it is temporary fix and we will get rid of the STI soon.

**Review by tbrisker — APPROVED** [review]
Thanks @ezr-ondrej ! Looks like the best option for now until we can drop the old definitions.

**Comment by xprazak2** [comment]
My settings are still reloading, and it is very annoying:

settings-reload

Is there anything I need to do for this fix to work?

**Comment by tbrisker** [comment]
2.5 - 6942751d5b236234f09b918e5b295e916d50e378

**Pull request merged**